WallarooLabs/wallaroo
# Revive spike

_Note_: this PR does **not** fix reconnection on a disconnected boundary. That should be addressed in its own issue.

As a result, you can only manually test that spike is working by running any wallaroo application with spike and observing that it does indeed drop the boundary's TCP connection.

## Changes:
- Add spike config to wallaroo_config
- Add updated spike
- Add SpikeConfig class to hold spike parameters
- Add new methods to NullWallarooOutgoingNetworkActor
- Update spike tests
- Update spike's drop_connection wrapper
- Use fraction for probabilities and set default to 0.001
- Add setting for minimum margin between spikes, set default to 10
- Add boudnary-reconnect playbook
- Fix off-by-1 seq_id deduplication
- Fix typo in printf: 'control' -> 'data'
- Replace env.out.print with @printf where encountered
- Plug spike all the way through from startup

## Testing:
With the `sequence-window` application, you can verify spike works by taking the following steps:

1. build sequence-window with spike: `stable env ponyc -d -D spike -D spiketrace`
1. run receiver: `../../../../giles/receiver/receiver --ponythreads=1 --ponynoblock --ponypinasio -l 127.0.0.1:5555`
1. run initializer, setting spike-probability to 1 and margin to 100: `./sequence-window -i 127.0.0.1:7000 -o 127.0.0.1:5555 -m 127.0.0.1:5001 --ponythreads=4 --ponypinasio --ponynoblock -c 127.0.0.1:12500 -d 127.0.0.1:12501 -r res-data -w 2 -n worker1 -t --spike-drop --spike-prob 1 --spike-margin 100`
1. run second worker with spike off (so only initializer drops): `./sequence-window -i 127.0.0.1:7000 -o 127.0.0.1:5555 -m 127.0.0.1:5001 --ponythreads=4 --ponypinasio --ponynoblock -c 127.0.0.1:12500 -d 127.0.0.1:12501 -r res-data -w 2 -n worker2`
1. send some data with sender: `../../../../giles/sender/sender -h 127.0.0.1:7000 -s 1 -i 50_000_000 --ponythreads=1 -y -g 12 -w -u -m 1000`

After a little under 200 messages (only half the input messages go over the boundary because of the partitioning), you should see the following in your process stdout:
```bash
output: [177,179,181,183]

<<<<<<SPIKE: DROPPING CONNECTION!>>>>>>

SPIKE Counts: {connecting: 1, sentv: 94, connected: 1, received: 4}
writev error: 32, iovcnt: 2
BoundaryNotify: closed

MUTE
```

And on the second worker:
```bash
output: [176,178,180,182]
DataChannelConnectNotifier : server closed
```


closes #335
